File: solutions_java/hashmaps/383.ransom-note.java
package solutions_java.hashmaps;
/*
 * @lc app=leetcode id=383 lang=java
 *
 * [383] Ransom Note
 */

// @lc code=start
class Solution {
    public boolean canConstruct(String ransomNote, String magazine) {
        Map<Character, Integer> map = new HashMap<>();
        for (int i=0; i<magazine.length(); i++) {
            if (!map.containsKey(magazine.charAt(i))) {
                map.put(magazine.charAt(i), 1);
            } else {
                map.put(magazine.charAt(i), map.get(magazine.charAt(i))+1);
            }
        }
        for (int i=0; i<ransomNote.length(); i++) {
          if (map.containsKey(ransomNote.charAt(i)) && map.get(ransomNote.charAt(i))>0){
            map.put(ransomNote.charAt(i), map.get(ransomNote.charAt(i))-1);
          } else {
            return false;
          }
        }
        return true;
    }
}
// @lc code=end

File: solutions/hashmaps/383.ransom-note.py
#
# @lc app=leetcode id=383 lang=python
#
# [383] Ransom Note
#

# @lc code=start
class Solution(object):
    def canConstruct(self, ransomNote, magazine):
        """
        :type ransomNote: str
        :type magazine: str
        :rtype: bool
        """
        mydict = {}

        for char in magazine:
            if char not in mydict:
                mydict[char] = 1
            else:
                mydict[char] += 1
        
        for char in ransomNote:
            if char not in mydict:
                return False
            else:
                mydict[char] -= 1
                if mydict[char] < 0:
                    return False
                
        return True

        
# @lc code=end

